File: mpp02_ephemeris.py
"""
Calculates position and velocity of the Moon w.r.t. the Earth using ELP/MPP02.
Can use LLR or DE405 fit data and their truncated versions.

ELP/MMP02 article: Lunar solution ELP, version ELP/MPP02, Jean Chapront and Gerard Francou. 
"""
import numpy as np

import tools.misc as misc

class MPP02Ephemeris:
    """
    Computes position and velocity of the Moon using ELP/MPP02.
    Units: km for position, km/Julian day for velocity.
    """
    ECL_TO_EQU = misc.rotation_matrix(0, 84381.448 * misc.ARCSEC)  # angle ~23.44 deg

    def __init__(self, source):
        """
        Source can be file path to json file or object loaded from the json file.
        """
        if isinstance(source, str):
            source = misc.load_json(source)
        self.W = source['W']
        self.PC = source['PC']
        self.QC = source['QC']
        self.groups = source['groups']
        # Convert coeffs arrays to numpy arrays.
        for group in self.groups:
            group['coeffs'] = np.array(group['coeffs']).reshape(-1, 6)

    def get_pos_vel(self, t: float):
        """
        Computes position and velocity.
        """
        t_pow = np.power(t, np.arange(6))
        t_pow_p = np.array([k * t_pow[k-1] if k > 0 else 0.0 for k in range(6)])

        v = np.zeros(3)
        vp = np.zeros(3)
        # Series terms have form c0 * t^\alpha * sin(c1 + c2*t + c3*t^2 + c4*t^3 + c5*t^4)
        for group in self.groups:
            coord = group['coord']
            alpha = group['alpha']
            coeffs = group['coeffs']

            a = coeffs[:,1].copy()
            ap = np.zeros(shape=a.shape)
            for k in range(1, 5):
                a += t_pow[k] * coeffs[:,k+1]
                ap += (k * t_pow[k-1]) * coeffs[:,k+1]
            c0_sin = np.sum(coeffs[:,0] * np.sin(a))
            c0_sin_p = np.sum(coeffs[:,0] * ap * np.cos(a))
            v[coord] += t_pow[alpha] * c0_sin
            vp[coord] += t_pow_p[alpha] * c0_sin + t_pow[alpha] * c0_sin_p

        # Compute spherical coordinates and their derivatives for the mean ecliptic of date.
        v[0] = v[0]*misc.ARCSEC + np.sum(self.W * t_pow[:5])
        v[1] *= misc.ARCSEC
        v[2] *= 0.9999999498265191
        vp[0] = vp[0]*misc.ARCSEC + np.sum(self.W * t_pow_p[:5])
        vp[1] *= misc.ARCSEC
        vp[2] *= 0.9999999498265191

        # Change to cartesian coordinates
        h = np.array([
            v[2]*np.cos(v[1])*np.cos(v[0]), 
            v[2]*np.cos(v[1])*np.sin(v[0]), 
            v[2]*np.sin(v[1])
        ])
        hp = np.array([
            (vp[2]*np.cos(v[1]) - vp[1]*v[2]*np.sin(v[1]))*np.cos(v[0]) - vp[0]*h[1],
            (vp[2]*np.cos(v[1]) - vp[1]*v[2]*np.sin(v[1]))*np.sin(v[0]) + vp[0]*h[0],
            vp[2]*np.sin(v[1]) + vp[1]*v[2]*np.cos(v[1])
        ])

        # Rotate from mean ecliptic and equinox of date to mean ecliptic and equinox of J2000.0.
        p = np.sum(self.PC * t_pow)
        q = np.sum(self.QC * t_pow)
        pp = np.sum(self.PC * t_pow_p)
        qp = np.sum(self.QC * t_pow_p)

        sc = np.sqrt(1.0 - p*p - q*q)
        pc1 = 1.0 - 2.0*p*p
        qc1 = 1.0 - 2.0*q*q
        pqp = pp*q + p*qp
        d2p = 2.0*p*pp + 2.0*q*qp
        pc2 = pp*sc - p*d2p/sc
        qc2 = qp*sc - q*d2p/sc

        pos = np.zeros(3)
        pos[0] = pc1*h[0] + 2.0*p*q*h[1] + 2.0*p*sc*h[2]
        pos[1] = 2.0*p*q*h[0] + qc1*h[1] - 2.0*q*sc*h[2]
        pos[2] = -2.0*p*sc*h[0] + 2.0*q*sc*h[1] + (pc1+qc1-1.0)*h[2]
        
        vel = np.zeros(3)
        vel[0] = pc1*hp[0] + 2.0*p*q*hp[1] + 2.0*p*sc*hp[2] 
        vel[0] += -4.0*p*pp*h[0] + 2.0*pqp*h[1] + 2.0*pc2*h[2]
        vel[1] = 2.0*p*q*hp[0] + qc1*hp[1] - 2.0*q*sc*hp[2] 
        vel[1] += 2.0*pqp*h[0] - 4.0*q*qp*h[1] - 2.0*qc2*h[2]
        vel[2] = -2.0*p*sc*hp[0] + 2.0*q*sc*hp[1] + (pc1+qc1-1.0)*hp[2] 
        vel[2] += -2.0*pc2*h[0] + 2.0*qc2*h[1] - 2.0*d2p*h[2]

        # Rotate from mean ecliptic and equinox of J2000.0 to mean equator and equinox of J2000.0.
        pos_equatorial = self.ECL_TO_EQU @ pos
        vel_equatorial = self.ECL_TO_EQU @ vel

        return pos_equatorial, vel_equatorial / 36525.0

def _test():
    # mpp02 = MPP02Ephemeris('./json/mpp02_llr_raw.json')
    mpp02 = MPP02Ephemeris('./json/mpp02_llr_truncated_medium.json')
    p, v = mpp02.get_pos_vel(-5.25)
    print(p.tolist(), v.tolist())

if __name__ == '__main__':
    _test()

File: javascript/mpp02Ephemeris.js
const { rotationMatrix, matrixMult, vectorMult } = require('./tools');

/**
 * Computes position and velocity of the Moon using ELP/MPP02.
 * Units: km for position, km/Julian day for velocity.
 */
class MPP02Ephemeris {
    PC = new Float64Array([0, 0.10180391e-4, 0.47020439e-6, -0.5417367e-9, -0.2507948e-11, 0.463486e-14]);
    QC = new Float64Array([0, -0.113469002e-3, 0.12372674e-6, 0.1265417e-8, -0.1371808e-11, -0.320334e-14]);
    ARCSEC = Math.PI / 180 / 3600;
    ECL_TO_EQU = rotationMatrix(0, 84381.448 * this.ARCSEC);    // angle ~23.44 deg

    /**
     * Here source has to be object containing the (possibly truncated) ELP/MPP02 series
     * loaded from one of the json files.
     */
    constructor(source) {
        this.W = source['W'];
        this.groups = source['groups'];
    }

    /**
     * Returns position and velocity for the Moon given time in Julian centuries since J2000.0. 
     * Units are km, Julian day.
     */
    getPosVel(t) {
        const tPow = Array.from({ length: 6 }, (_, k) => Math.pow(t, k));
        const tPowP = Array.from({ length: 6 }, (_, k) => k > 0 ? k * tPow[k-1] : 0);

        let v = [0, 0, 0];
        let vp = [0, 0, 0];
        // Series terms have form c0 * t^\alpha * sin(c1 + c2*t + c3*t^2 + c4*t^3 + c5*t^4)
        this.groups.forEach((group) => {
            const coord = group['coord'];
            const alpha = group['alpha'];
            const coeffs = group['coeffs'];

            for (let k6 = 0; k6 < coeffs.length; k6 += 6) {
                let a = 0;
                let ap = 0;
                for (let j = 0; j < 5; j++) {
                    const c = coeffs[k6 + j + 1];
                    a += c * tPow[j];
                    ap += c * tPowP[j];
                }
                v[coord] += coeffs[k6 + 0] * tPow[alpha] * Math.sin(a);
                vp[coord] += coeffs[k6 + 0] * (tPowP[alpha] * Math.sin(a) + tPow[alpha] * ap * Math.cos(a));
            }
        });

        // Compute spherical coordinates and their derivatives for the mean ecliptic of date.
        v[0] = v[0]*this.ARCSEC + this.W.reduce((acc, value, k) => acc + value * tPow[k], 0);
        v[1] *= this.ARCSEC;
        v[2] *= 0.9999999498265191;
        vp[0] = vp[0]*this.ARCSEC + this.W.reduce((acc, value, k) => acc + value * tPowP[k], 0);
        vp[1] *= this.ARCSEC;
        vp[2] *= 0.9999999498265191;

        // Change to cartesian coordinates
        const h = [
            v[2]*Math.cos(v[1])*Math.cos(v[0]), 
            v[2]*Math.cos(v[1])*Math.sin(v[0]), 
            v[2]*Math.sin(v[1])
        ];
        const hp = [
            (vp[2]*Math.cos(v[1]) - vp[1]*v[2]*Math.sin(v[1]))*Math.cos(v[0]) - vp[0]*h[1],
            (vp[2]*Math.cos(v[1]) - vp[1]*v[2]*Math.sin(v[1]))*Math.sin(v[0]) + vp[0]*h[0],
            vp[2]*Math.sin(v[1]) + vp[1]*v[2]*Math.cos(v[1])
        ];

        // Next, rotate from mean ecliptic and equinox of date to mean ecliptic and equinox of J2000.0.
        const p = this.PC.reduce((acc, value, k) => acc + value * tPow[k], 0);
        const q = this.QC.reduce((acc, value, k) => acc + value * tPow[k], 0);
        const pp = this.PC.reduce((acc, value, k) => acc + value * tPowP[k], 0);
        const qp = this.QC.reduce((acc, value, k) => acc + value * tPowP[k], 0);

        const sc = Math.sqrt(1 - p*p - q*q);
        const pc1 = 1 - 2*p*p;
        const qc1 = 1 - 2*q*q;
        const pqp = pp*q + p*qp;
        const d2p = 2*p*pp + 2*q*qp;
        const pc2 = pp*sc - p*d2p/sc;
        const qc2 = qp*sc - q*d2p/sc;

        const pos = [0, 0, 0];
        pos[0] = pc1*h[0] + 2*p*q*h[1] + 2*p*sc*h[2];
        pos[1] = 2*p*q*h[0] + qc1*h[1] - 2*q*sc*h[2];
        pos[2] = -2*p*sc*h[0] + 2*q*sc*h[1] + (pc1+qc1-1)*h[2];
        
        const vel = [0, 0, 0];
        vel[0] = pc1*hp[0] + 2*p*q*hp[1] + 2*p*sc*hp[2];
        vel[0] += -4*p*pp*h[0] + 2*pqp*h[1] + 2*pc2*h[2];
        vel[1] = 2*p*q*hp[0] + qc1*hp[1] - 2*q*sc*hp[2];
        vel[1] += 2*pqp*h[0] - 4*q*qp*h[1] - 2*qc2*h[2];
        vel[2] = -2*p*sc*hp[0] + 2*q*sc*hp[1] + (pc1+qc1-1)*hp[2];
        vel[2] += -2*pc2*h[0] + 2*qc2*h[1] - 2*d2p*h[2];

        // Finally, rotate from mean ecliptic and equinox of J2000.0 to mean equator and equinox of J2000.0.
        const posEquatorial = matrixMult(this.ECL_TO_EQU, pos);
        const velEquatorial = matrixMult(this.ECL_TO_EQU, vel);

        return { 'pos': posEquatorial, 'vel': vectorMult(velEquatorial, 1/36525) };
    }
}

module.exports = MPP02Ephemeris;